### ISSUE TITLE
ZLayer.wire macro fails on compile time

### ISSUE THREAD

**Issue by vladimirkl** [kitlangton/zio-magic#53]
I found an interesting case and extracted it from real project:
```
object Example {
  import zio._
  import zio.magic._

  type A = Has[A.Service]
  object A {
    trait Service {}
    val live: ULayer[A] = ZLayer.succeed(new Service {})
  }

  type B = Has[B.Service]
  object B {
    trait Service {}
    val live: ULayer[B] =
      ZLayer.succeed(new Service {})
  }

  type C = Has[C.Service]
  object C {
    trait Service {}
    val live: URLayer[A with B, C] =
      ZLayer.succeed(new Service {})
  }

  // Success
  val layer1: ULayer[A with B] = A.live >+> B.live >+> C.live

  // Failure
  val layer2: ULayer[A with B] = ZLayer.wire[A with B](A.live, B.live, C.live)
}

type mismatch;
 found   : zio.ZLayer[Example.A with Example.B with Any with Any,Nothing,Example.C with Example.A with Example.B]
    (which expands to)  zio.ZLayer[zio.Has[Example.A.Service] with zio.Has[Example.B.Service] with Any with Any,Nothing,zio.Has[Example.C.Service] with zio.Has[Example.A.Service] with zio.Has[Example.B.Service]]
 required: zio.ZLayer[Any,Nothing,zio.Has[Example.A.Service] with zio.Has[Example.B.Service]]
  val layer2: ULayer[A with B] = ZLayer.wire[A with B](A.live, B.live, C.live)
```

**Comment by vladimirkl** [comment]
This case should be supported, for example it's useful for mixing layer that runs some background process, but not directly used in program.

**Comment by kitlangton** [comment]
Ah, yes. @hmemcpy reached out to me with a similar issue.
I was leaning towards reporting a specific/readable error when you provide excess layers.
Though it's very useful to use a layer as a background process as you mentioned. However, maybe that's an opportunity for a more specific helper method on ZIO, a sort of managed `tap`.

**Comment by vladimirkl** [comment]
Please note that excess dependency works until it depends on others - so if we define it as `val live: ULayer[C]=???` it will compile. I think that automatic composition should not be less powerful than manual. Personally I use such things in integration tests - to refactor setup / cleanup of some external resource as layer and reuse between specs

**Comment by kitlangton** [comment]
Pushed a fix :) It should now build the extra, side-effecting layer as well as prompt you for its transitive dependencies. Publishing to Sonatype now. Let me know if it works for ya 🤗

**Comment by hmemcpy** [comment]
I confirm the latest fix is MAGICAL! It solves my issue (which is very similar to @vladimirkl's). So thank you :))

BTW, this should be fixed for ZIO 2.0, right?
![image](https://user-images.githubusercontent.com/601206/114515846-78e6cc00-9c45-11eb-93fe-691d26ad6a0d.png)
(the `package` string, it should be `sesv2.live`)

**Comment by kitlangton** [comment]
> I confirm the latest fix is MAGICAL! It solves my issue (which is very similar to @vladimirkl's). So thank you :))
> 
> BTW, this should be fixed for ZIO 2.0, right?
> ![image](https://user-images.githubusercontent.com/601206/114515846-78e6cc00-9c45-11eb-93fe-691d26ad6a0d.png)
> (the `package` string, it should be `sesv2.live`)

It should be. I have some tests for that. It’s surprisingly difficult to print the actual name in Scala 2. I’ll check that out :)

**Comment by vladimirkl** [comment]
Also works for me, thank you!
